File: CfxGenerator/ApiTypes/CefBasePtrType.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Public Class CefBasePtrType
    Inherits ApiType
    Sub New()
        MyBase.New("struct _cef_base_t*")
    End Sub

    Public Overrides ReadOnly Property PInvokeSymbol As String
        Get
            Return "IntPtr"
        End Get
    End Property

    Public Overrides ReadOnly Property PublicSymbol As String
        Get
            Return "CfxBase"
        End Get
    End Property

    Public Overrides ReadOnly Property ProxySymbol As String
        Get
            Return "IntPtr"
        End Get
    End Property

    Public Overrides Function PublicWrapExpression(var As String) As String
        Return String.Format("CfxBase.Cast({0})", var)
    End Function

    Public Overrides Function PublicUnwrapExpression(var As String) As String
        Return String.Format("CfxBase.Unwrap({0})", var)
    End Function

    Public Overrides Function ProxyWrapExpression(var As String) As String
        Return String.Format("CfxBase.Unwrap({0})", var)
    End Function

    Public Overrides Function ProxyUnwrapExpression(var As String) As String
        Return String.Format("CfxBase.Cast({0})", var)
    End Function

End Class

File: CfxGenerator/ApiTypes/CefBasePtrType.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

public class CefBaseRefCountedPtrType : ApiType {

    public CefBaseRefCountedPtrType()
        : base("struct _cef_base_ref_counted_t*") {
    }

    public override string PInvokeSymbol {
        get { return "IntPtr"; }
    }

    public override string PublicSymbol {
        get { return "CfxBaseRefCounted"; }
    }

    public override string ProxySymbol {
        get { return "IntPtr"; }
    }

    public override string RemoteSymbol {
        get { return "CfrBaseRefCounted"; }
    }

    public override string[] ParserMatches {
        get {
            return new string[] {
                Name,
                "cef_base_ref_counted_t*"
            };
        }
    }

    public override void EmitPreRemoteCallStatements(CodeBuilder b, string var) {
        b.AppendLine("if(!CfrObject.CheckConnection({0}, connection)) throw new ArgumentException(\"Render process connection mismatch.\", \"{1}\");", CSharp.Escape(var), var);
        b.AppendLine("call.{0} = CfrBaseRefCounted.Unwrap({0}).ptr;", CSharp.Escape(var));
    }


    public override string PublicWrapExpression(string var) {
        return string.Format("CfxBaseRefCounted.Cast({0})", var);
    }

    public override string PublicUnwrapExpression(string var) {
        return string.Format("CfxBaseRefCounted.Unwrap({0})", var);
    }

    public override string ProxyWrapExpression(string var) {
        return string.Format("CfxBaseRefCounted.Unwrap({0})", var);
    }

    public override string ProxyUnwrapExpression(string var) {
        return string.Format("CfxBaseRefCounted.Cast({0})", var);
    }

    public override string RemoteWrapExpression(string var) {
        return string.Format("CfrBaseRefCounted.Cast(new RemotePtr(connection, {0}))", var);
    }

    public override string ProxyReturnExpression(string var) {
        return var;
    }

    public override string ProxyCallArgument(string var) {
        return var;
    }
}